src/components/Firestore/Field/index.tsx

```diff
@@ -14,37 +14,206 @@
-import { useDocumentRef } from './DocumentRefContext';
+import * as actions from './actions';
+import { FirestoreAny } from './models';
+import {
+  useFieldState,
+  useDocumentDispatch,
+  DocumentProvider,
+} from './DocumentStore';
+import { Editor } from './Editor';
+import {
+  isMap,
+  isArray,
+  isPrimitive,
+  getFieldType,
+  getLeafPath,
+} from './utils';
 
 import './index.scss';
 
-export const Field: React.FC<{
-  id: string;
-  value: any;
-}> = ({ id, value }) => {
-  const reference = useDocumentRef();
+/** Entry point for a Document/Field preview */
+export const Field: React.FC<{ value: FirestoreAny }> = ({ value }) => {
+  return (
+    <DocumentProvider value={value}>
+      <RootField />
+    </DocumentProvider>
+  );
+};
 
-  function handleDeleteField() {
-    reference.update({ [id]: firestore.FieldValue.delete() });
+/**
+ * Special representation of a Document Root, where we don't want to show
+ * the implicit top-level map.
+ */
+const RootField: React.FC = () => {
+  const state = useFieldState([]);
+  return (
+    <>
+      {isMap(state) &&
+        Object.keys(state).map(name => (
+          <NestedField key={name} path={[name]} />
+        ))}
+    </>
+  );
+};
+
+/**
+ * Field with CTAs for editing as well as rendering applicable child-nodes
+ */
+const NestedField: React.FC<{
+  path: string[];
+  edit?: boolean;
+}> = ({ path, edit = false }) => {
+  const state = useFieldState(path);
+  const dispatch = useDocumentDispatch()!;
+  const [isExpanded, setIsExpanded] = useState(true);
+  const [isAddingField, setIsAddingField] = useState(false);
+  const [isEditing, setIsEditing] = useState(edit);
+
+  const fieldName = getLeafPath(path);
+
+  function handleDeleteField(e: React.MouseEvent<HTMLButtonElement>) {
+    e.stopPropagation();
+    dispatch(actions.deleteField(path));
   }
 
-  return (
-    <ListItem>
-      {id}: {JSON.stringify(value)}
-      <ListItemMeta className="Firestore-Field-actions">
-        ({getFieldType(value)})
-        <IconButton
-          icon="delete"
-          label="Remove field"
-          onClick={handleDeleteField}
+  function handleEditField(e: React.MouseEvent<HTMLButtonElement>) {
+    e.stopPropagation();
+    setIsEditing(true);
+  }
+
+  function handleEditCancel() {
+    setIsEditing(false);
+  }
+
+  function handleEditSave(value: FirestoreAny) {
+    dispatch(actions.updateField({ path, value }));
+    setIsEditing(false);
+  }
+
+  function handleExpandToggle() {
+    setIsExpanded(!isExpanded);
+  }
+
+  function handleAddField(e: React.MouseEvent<HTMLButtonElement>) {
+    e.stopPropagation();
+    setIsAddingField(true);
+  }
+
+  function handleAddCancel() {
+    setIsAddingField(false);
+  }
+
+  function handleAddSave(value: FirestoreAny) {
+    dispatch(actions.addField({ path: addPath, value }));
+    setIsAddingField(false);
+  }
+
+  let addPath: string[] = [];
+  let childFields = null;
+  if (isMap(state)) {
+    addPath = [...path, ''];
```

Comment:
Is this `""` intentional? If so, why do we need this?

Reply:
Yeah, added comment inline. This is the default key we pass into the child-editor.